bloom.go

```diff
@@ -0,0 +1,89 @@
+// Package bloom implements a Bloom filter data structure.
+package bloom
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"log"
+	"math/bits"
+)
+
+// Filter represents a Bloom filter. Filter satisfies the encoding.BinaryMarshaler and BinaryUnmarshaler interfaces.
+// The zero value represents an empty filter of size 0 that uses 0 hash values.
+// Current implementations of Insert and Contains require len(f) <= 2^16 bits and k <= 16.
+type Filter struct {
+	f []byte
+	k int
+}
+
+// bit returns the filter's nth bit.
+func (f *Filter) bit(n int) int {
+	b, i := n/8, n%8
+	return int(f.f[b]>>uint(i)) & 1
+}
+
+// setBit sets the filter's nth bit to 1.
+func (f *Filter) setBit(n int) {
+	b, i := n/8, n%8
+	f.f[b] |= 1 << uint(i)
+}
+
+// hashBits returns the base 2 logarithm of the size of the filter in bits.
+// This is the number of hash bits required to index any bit in the filter.
+// TODO: Handle filter size not a power of 2
+func (f *Filter) hashBits() int {
+	if bits.OnesCount(uint(len(f.f))) != 1 {
+		log.Fatal("hashBits: filter size not a power of 2")
+	}
+	return bits.TrailingZeros64(uint64(len(f.f)) * 8)
+}
+
+// New returns a Filter of size b bytes that uses k hash values.
+func New(b, k int) *Filter { return &Filter{make([]byte, b), k} }
```

Comment:
I strongly suggest checking size requirements explicitly here, and panic immediately if they aren't met, rather than letting the process go on for a while and then (presumably?) panicking later when the client package tries to use the filter. If it doesn't panic later and just gives incorrect results, then even more important to check for them here.

If the size also has to be a power of 2, it should check for that here as well.